### PR TITLE
Preserve locale through startJVM

### LINKED ISSUE #1078: Bug: datetime.strptime not working correctly after calling startJVM
## Bug description

Hello, I found a weird behavior of the standard Python `datetime` module after starting up the JVM. Namely, I found out that using `strptime` with a format containing the `%B` modifier does not work.

## Steps to reproduce the behavior

```python
import jpype
import datetime

# This line works correctly -> 2020-10-27 00:00:00
print(datetime.datetime.strptime("27 October 2020", "%d %B %Y"))

jpype.startJVM()

# This line (which is the same as the previous one) throws
# ValueError: time data '27 October 2020' does not match format '%d %B %Y'
print(datetime.datetime.strptime("27 October 2020", "%d %B %Y"))
```

## Expected behavior

In both cases, I should get the correct datetime without any exception.

## Environment

```
Python 3.9.7
JPype1 1.4.0
```

### PR DESCRIPTION
**Pull request by Thrameos**
Patch to fix behavior with locale.   It appears that Python and Java are using the same C library for locale.  When starting the JVM it reinitialized the locale to different values than the current Python settings.  To fix this we store the locale variables prior to starting the JVM and restore them so that we don't mess up the user settings.

It is not clear if this needs to be a switchable behavior.  If it should be optional then I would recommend preserving by default as it is the expected behavior. 

Fixes #1078

### CODE REVIEW

**Review by marscher — APPROVED** [review]
not sure if the list of LC vars is complete, bit looks like a sane approach. Could we reset everything which java forcefully sets for us as well?

**Comment by Thrameos** [comment]
I tried a few other approaches rather than listing the values to restore explicitly, but unfortunately values like LC_ALL do not work for get and produced an error.   Worse LC_ALL is the highest ordinal on linux and the lowest ordinal on windows so I couldn't just iterate through them by number.   LC_MESSAGES is missing on windows and appears on linux.   So basically trying to restore and preserve is a mess.   So I went simple and did the lowest common denominator.   However, that also may not be adequate as some other system may be missing one of the LC.

An alternative approach would be to scan the dictionary of locale and find all LC excluding LC_ALL, save them, start the JVM, then restore.  I was not able to find a call in locale which would give me a list of all the settable categories.

With regard to the last comment "Could we reset everything which java forcefully sets for us as well?"  I am not sure what other settings that Java is forcing on the user.   With the exception of the locale which is clearly a common library, do you know of any other settings that Python and Java are sharing?

**Comment by Thrameos** [comment]
Would this line look too hackish?

``` 
  pre = [locale.getlocale(getattr(locale, i)) for i in locale.__all__ if i.startswith('LC_') and i!='LC_ALL']
```

**Comment by marscher** [comment]
> An alternative approach would be to scan the dictionary of locale and find all LC excluding LC_ALL, save them, start the JVM, then restore. I was not able to find a call in locale which would give me a list of all the settable categories.

thats what i had in mind.

> Would this line look too hackish?

no, i u put a comment for it.

> do you know of any other settings that Python and Java are sharing?

unfortunately not

**Review by marscher — APPROVED** [review]

File: jpype/_core.py [hunk 1]
```diff
@@ -215,9 +215,19 @@ def startJVM(*args, **kwargs):
                         % (','.join([str(i) for i in kwargs])))
 
     try:
+        import locale
+        # Gather a list of locale settings that Java may override (excluding LC_ALL)
+        categories = [getattr(locale, i) for i in locale.__all__ if i.startswith('LC_') and i != 'LC_ALL']
```
Comment (pelson) [inline jpype/_core.py:220]: Accessing ``__all__`` is fragile (https://docs.python.org/3/tutorial/modules.html?highlight=__all__#importing-from-a-package), even if in practice ``locale`` always has it.

Rather to use ``dir`` to access the attributes of ``locale``:

```
>>> print(', '.join(sorted(name for name in locale.__all__ if name.startswith('LC_'))))
LC_ALL, LC_COLLATE, LC_CTYPE, LC_MESSAGES, LC_MONETARY, LC_NUMERIC, LC_TIME

>>> print(', '.join(name for name in sorted(dir(locale)) if name.startswith('LC_')))
LC_ALL, LC_COLLATE, LC_CTYPE, LC_MESSAGES, LC_MONETARY, LC_NUMERIC, LC_TIME
```

Or, if you want to get the attribute values, we can use ``vars``:

```
categories = [v for k, v in vars(locale).items() if k.startswith('LC_')]
```

**Comment by pelson** [comment]
I saw this because it fails one of my CI pipelines (which is tested against JPype master for this reason). Pipeline may (possibly) be seen at https://gitlab.cern.ch/scripting-tools/pyjapc/-/jobs/22892411, but in essence:

```
______________________ ERROR at setup of test_cause_chain ______________________
    @pytest.fixture(scope="module")
    def jvm():
        mgr = cmmnbuild_dep_manager.Manager('pyjapc')
>       mgr.start_jpype_jvm()
/opt/conda/envs/test_env/lib/python3.10/site-packages/pyjapc/tests/conftest.py:76: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/conda/envs/test_env/lib/python3.10/site-packages/cmmnbuild_dep_manager/_manager.py:319: in start_jpype_jvm
    self.JVM.required()
/opt/conda/envs/test_env/lib/python3.10/site-packages/cmmnbuild_dep_manager/jvm/_manager.py:50: in required
    jp.startJVM(*jvm_args, convertStrings=True)
/opt/conda/envs/test_env/lib/python3.10/site-packages/jpype/_core.py:230: in startJVM
    locale.setlocale(i, j)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
category = 0, locale = 'en_US.UTF-8'
    def setlocale(category, locale=None):
    
        """ Set the locale for the given category.  The locale can be
            a string, an iterable of two strings (language code and encoding),
            or None.
    
            Iterables are converted to strings using the locale aliasing
            engine.  Locale strings are passed directly to the C lib.
    
            category may be given as one of the LC_* values.
    
        """
        if locale and not isinstance(locale, _builtin_str):
            # convert to string
            locale = normalize(_build_localename(locale))
>       return _setlocale(category, locale)
E       locale.Error: unsupported locale setting
/opt/conda/envs/test_env/lib/python3.10/locale.py:620: Error
```

This test was passing before the change here. It is completely believable that the locale value doesn't make any sense (didn't check), but wanted to let you know ASAP about the potential for breakage with this PR.

**Comment by marscher** [comment]
one could also handle this kind of exception, right?

**Comment by pelson** [comment]
Yep, it is easy to reproduce:

```
>>> for l in [ getattr(locale,i) for i in dir(locale) if i.startswith('LC_') and i!='LC_ALL']:
...     print(l, locale.getlocale(l))
...     locale.setlocale(l, locale.getlocale(l))
... 
3 (None, None)
'C'
0 ('en_US', 'UTF-8')
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
  File "/opt/conda/lib/python3.9/locale.py", line 610, in setlocale
    return _setlocale(category, locale)
locale.Error: unsupported locale setting

```

This can be done on Python from ``docker run -it continuumio/miniconda3``. I certainly feel that this is a bug with that image (will report shortly), but also worth noting that the change here introduces new fail modes.


Handling this exception works fine, FWIW:

```
for l in [ getattr(locale,i) for i in dir(locale) if i.startswith('LC_') and i!='LC_ALL']:
    try:
        locale.setlocale(l, locale.getlocale(l))
    except locale.Error:
        pass
```

**Comment by Thrameos** [comment]
Thanks.  I applied the fix in #1080.

**Comment by pelson** [comment]
Confirm that #1080 fixes it properly :+1:

**Pull request merged**